File: ruby/lib/mpp/protocol/core/credential.rb
# frozen_string_literal: true

require "pay_core/base64_url"
require "pay_core/json"

module Mpp
  module Protocol
    module Core
      # Payment credential carried by the `Authorization` header.
      class Credential
        MAX_TOKEN_LENGTH = 16 * 1024

        attr_reader :challenge, :payload, :source

        def initialize(challenge:, payload:, source: nil)
          raise ArgumentError, "payload must be an object" unless payload.is_a?(Hash)

          @challenge = challenge
          @payload = payload
          @source = source
        end

        # Serialize to the wire credential shape.
        def to_h
          value = {
            "challenge" => challenge.to_h,
            "payload" => payload
          }
          value["source"] = source unless source.nil?
          value
        end

        # Format as `Authorization: Payment ...` value.
        def to_authorization_header
          "Payment #{::PayCore::Base64Url.encode(::PayCore::Json.canonical_generate(to_h))}"
        end

        # Parse an `Authorization` header value.
        def self.from_authorization_header(header)
          token = extract_payment_token(header)
          raise ArgumentError, "expected Payment scheme" if token.nil?
          raise ArgumentError, "token exceeds maximum length" if token.bytesize > MAX_TOKEN_LENGTH

          decoded = ::PayCore::Json.parse(::PayCore::Base64Url.decode(token))
          new(
            challenge: ChallengeEcho.from_h(decoded.fetch("challenge")),
            payload: decoded.fetch("payload"),
            source: decoded["source"]
          )
        end

        def self.extract_payment_token(header)
          header.to_s.split(",").map(&:strip).find { |part| part.downcase.start_with?("payment ") }&.[](8..)&.strip
        end
      end
    end
  end
end

File: php/src/Protocols/Mpp/Core/Credential.php
<?php

declare(strict_types=1);

namespace PayKit\Protocols\Mpp\Core;

use InvalidArgumentException;

/**
 * Represents the Payment authorization credential sent by a client.
 */
final class Credential
{
    /**
     * Create a credential that echoes a challenge and carries payment payload data.
     *
     * @param array<string, mixed> $payload
     */
    public function __construct(
        public readonly ChallengeEcho $challenge,
        public readonly array $payload = [],
        public readonly ?string $source = null,
    ) {
    }

    /**
     * Convert the credential to its JSON wire shape.
     *
     * @return array<string, mixed>
     */
    public function toArray(): array
    {
        $value = [
            'challenge' => $this->challenge->toArray(),
            'payload' => $this->payload,
        ];
        if ($this->source !== null) {
            $value['source'] = $this->source;
        }

        return $value;
    }

    /**
     * Format the credential as an Authorization header value.
     */
    public function toAuthorizationHeader(): string
    {
        return 'Payment ' . Base64Url::encodeJson($this->toArray());
    }

    /**
     * Parse an Authorization header into a Payment credential.
     */
    public static function fromAuthorizationHeader(string $header): self
    {
        $token = self::extractPaymentToken($header);
        if (strlen($token) > 16 * 1024) {
            throw new InvalidArgumentException('Token exceeds maximum length of 16384 bytes');
        }

        $decoded = Base64Url::decodeJson($token);
        if (!isset($decoded['challenge']) || !is_array($decoded['challenge'])) {
            throw new InvalidArgumentException('Invalid credential JSON structure');
        }

        $payload = $decoded['payload'] ?? [];
        if (!is_array($payload)) {
            throw new InvalidArgumentException('Credential payload must be an object');
        }

        return new self(
            challenge: ChallengeEcho::fromArray(Json::object($decoded['challenge'], 'challenge')),
            payload: Json::object($payload, 'payload'),
            source: isset($decoded['source']) ? Json::string($decoded['source'], 'source') : null,
        );
    }

    private static function extractPaymentToken(string $header): string
    {
        foreach (explode(',', $header) as $part) {
            $part = trim($part);
            if (stripos($part, 'Payment ') === 0) {
                return trim(substr($part, strlen('Payment ')));
            }
        }

        throw new InvalidArgumentException('Expected Payment scheme');
    }
}
